odoo/sql_db.py

```diff
@@ -300,10 +297,15 @@ def execute(self, query, params=None, log_exceptions=None):
             if self._default_log_exceptions if log_exceptions is None else log_exceptions:
                 _logger.error("bad query: %s\nERROR: %s", tools.ustr(self._obj.query or query), e)
             raise
+        finally:
+            delay = (time.time() - now)
+
+            if self.sql_log:
+                encoding = psycopg2.extensions.encodings[self.connection.encoding]
+                _logger.debug("query [%s]: %s", delay, self._obj.mogrify(query, params).decode(encoding, 'replace'))
```

Comment:
Is there a point in doing that in case of exception? Seems like this should be inside the `try`, no?

Also the downside of this change is that you don't see the query until it completes, which may be a very long time if you're debugging a performance issue. And we don't want to double the number of log lines just to log the time.

So it's a trade-off, but is it worth it, considering that you get the total SQL time already at the end of the request? Perhaps we should we have a separate logger for this case (e.g. `odoo.sql_db.times:DEBUG`), so devs can choose between the early logging, the SQL times, or both?


PS: whenever possible please include a proper description (rationale) and reference (OPW/task?) in your PR descriptions and commit messages. Even when it seems trivial, it's always useful to document your thought process for your colleagues (and for yourself later ;-))

Reply:
Having the new log line in the "finally" clause could help identify queries that fail slowly ? 
I'm not sure if that's a realistic use case.

Not having the early logging will definitely bug me.

Reply:
Would also add: `query [135]: select ...` seems rather unclear, numbers within brackets tend to be things like ids or identifiers (which this could be e.g. we could imagine logging extended details for every query). That this would likely be `query [0.0195069351196289] ...` does not seem singularly more helpful, possibly the opposite.

Reply:
I improved it by printing query number before and after the execution. Commit description is added ;-)

```
2021-02-05 09:12:19,967 49 DEBUG devel odoo.sql_db: query #1-167 started: SELECT "res_partner".id FROM "res_partner" WHERE ("res_partner"."id" in (3,2)) AND (((("res_partner"."partner_share" IS NULL or "res_partner"."partner_share" = false ) OR ("res_partner"."company_id" in (1))) OR "res_partner"."company_id" IS NULL ) AND ((("res_partner"."type" != 'private') OR "res_partner"."type" IS NULL) OR "res_partner"."type" IS NULL )) ORDER BY  "res_partner"."display_name"   
2021-02-05 09:12:19,968 49 DEBUG devel odoo.sql_db: query #1-167 ended: 0:00:00.000570 
2021-02-05 09:12:19,970 49 DEBUG devel odoo.sql_db: query #2-167 started: SELECT "res_partner"."id" as "id", "res_partner"."name" as "name", "res_partner"."display_name" as "display_name", "res_partner"."date" as "date", "res_partner"."title" as "title", "res_partner"."parent_id" as "parent_id", "res_partner"."ref" as "ref", "res_partner"."lang" as "lang", "res_partner"."tz" as "tz", "res_partner"."vat" as "vat", "res_partner"."website" as "website", "res_partner"."comment" as "comment", "res_partner"."credit_limit" as "credit_limit", "res_partner"."active" as "active", "res_partner"."employee" as "employee", "res_partner"."function" as "function", "res_partner"."type" as "type", "res_partner"."street" as "street", "res_partner"."street2" as "street2", "res_partner"."zip" as "zip", "res_partner"."city" as "city", "res_partner"."state_id" as "state_id", "res_partner"."country_id" as "country_id", "res_partner"."partner_latitude" as "partner_latitude", "res_partner"."partner_longitude" as "partner_longitude", "res_partner"."mobile" as "mobile", "res_partner"."is_company" as "is_company", "res_partner"."industry_id" as "industry_id", "res_partner"."company_id" as "company_id", "res_partner"."color" as "color", "res_partner"."partner_share" as "partner_share", "res_partner"."commercial_partner_id" as "commercial_partner_id", "res_partner"."commercial_company_name" as "commercial_company_name", "res_partner"."company_name" as "company_name", "res_partner"."message_main_attachment_id" as "message_main_attachment_id", "res_partner"."email_normalized" as "email_normalized", "res_partner"."message_bounce" as "message_bounce", "res_partner"."email" as "email", "res_partner"."phone" as "phone", "res_partner"."user_id" as "user_id", "res_partner"."contact_address_complete" as "contact_address_complete", "res_partner"."signup_token" as "signup_token", "res_partner"."signup_type" as "signup_type", "res_partner"."signup_expiration" as "signup_expiration", "res_partner"."phone_sanitized" as "phone_sanitized", "res_partner"."ocn_token" as "ocn_token", "res_partner"."partner_gid" as "partner_gid", "res_partner"."additional_info" as "additional_info", "res_partner"."create_uid" as "create_uid", "res_partner"."create_date" as "create_date", "res_partner"."write_uid" as "write_uid", "res_partner"."write_date" as "write_date" FROM "res_partner" WHERE "res_partner".id IN (2, 3) 
2021-02-05 09:12:19,971 49 DEBUG devel odoo.sql_db: query #2-167 ended: 0:00:00.000650 
2021-02-05 09:12:19,972 49 DEBUG devel odoo.sql_db: query #3-167 started: 
            SELECT
                U.partner_id as id,
                CASE WHEN max(B.last_poll) IS NULL THEN 'offline'
                    WHEN age(now() AT TIME ZONE 'UTC', max(B.last_poll)) > interval '55 seconds' THEN 'offline'
                    WHEN age(now() AT TIME ZONE 'UTC', max(B.last_presence)) > interval '1800 seconds' THEN 'away'
                    ELSE 'online'
                END as status
            FROM bus_presence B
            RIGHT JOIN res_users U ON B.user_id = U.id
            WHERE U.partner_id IN (2, 3) AND U.active = 't'
         GROUP BY U.partner_id
         
2021-02-05 09:12:19,973 49 DEBUG devel odoo.sql_db: query #3-167 ended: 0:00:00.001004 
2021-02-05 09:12:19,973 49 DEBUG devel odoo.sql_db: query #4-167 started: SELECT id, model, res_id FROM ir_model_data WHERE module='base' AND name='partner_root' 
2021-02-05 09:12:19,973 49 DEBUG devel odoo.sql_db: query #4-167 ended: 0:00:00.000315 
2021-02-05 09:12:19,974 49 DEBUG devel odoo.sql_db: SQL LOG from: 
2021-02-05 09:12:19,975 49 DEBUG devel odoo.sql_db: table: ir_model_data: 0:00:00.000315/1 
2021-02-05 09:12:19,975 49 DEBUG devel odoo.sql_db: table: res_partner: 0:00:00.001220/2 
2021-02-05 09:12:19,975 49 DEBUG devel odoo.sql_db: SUM from:0:00:00.001534/4 [171] 
2021-02-05 09:12:19,975 49 DEBUG devel odoo.sql_db: SUM into:0:00:00/4 [171] 
```

Reply:
I realized that the regexes used to detect the `SELECT` and `INSERT` queries are plain wrong:
* they are not raw strings (bad practice, but not a bug as such)
* they are case-sensitive: `from` matches, while `FROM` does not
* they are not multiline: most queries in the ORM start with a newline

Reply:
> I realized that the regexes used to detect the `SELECT` and `INSERT` queries are plain wrong:
> 
>     * they are not raw strings (bad practice, but not a bug as such)
>     * they are case-sensitive: `from` matches, while `FROM` does not
>     * they are not multiline: most queries in the ORM start with a newline

They have other issues e.g. `insert into foo (bar) select baz from qux` will match both filters.

Maybe the entire thing could be simplified to a simple predicate in order to differentiate readonly queries from any sort of write query? We may want to handle `WITH [RECURSIVE]` properly but otherwise it's way more straightforward: just check if the query starts with a `SELECT`.

Reply:
> they are case-sensitive: from matches, while FROM does not

they are case-sesitive, but are used after lower()

https://github.com/odoo/odoo/blob/7948802fa11ca5444b5da6bcfb9827af3c9a0bf4/odoo/sql_db.py#L317-L318